compositor_render/src/render_loop.rs

```diff
@@ -123,48 +126,60 @@ pub(super) fn run_transforms(
     scene: &mut Scene,
     pts: Duration,
 ) -> Result<(), SceneError> {
-    let node_id_iter = render_order_iter(scene)?;
-    for node_id in node_id_iter {
-        let NodeRenderPass { node, inputs } = scene.nodes.node_render_pass(&node_id)?;
-        let input_textures: Vec<_> = inputs
-            .iter()
-            .map(|node| (&node.node_id, &node.output))
-            .collect();
-        node.renderer
-            .render(ctx, &input_textures, &mut node.output, pts)
-    }
-    Ok(())
-}
-
-pub struct NodeRenderPass<'a> {
-    pub node: &'a mut Node,
-    pub inputs: Vec<&'a Node>,
-}
-
-/// Returns iterator that guarantees that nodes will be visited in
-/// the order that ensures that inputs of the current node were
-/// already visited in previous iterations
-fn render_order_iter(scene: &Scene) -> Result<impl Iterator<Item = NodeId>, SceneError> {
-    let mut queue = vec![];
+    let mut already_rendered = HashSet::new();
     for (node_id, _) in scene.outputs.values() {
-        enqueue_nodes(node_id, &mut queue, scene)?;
+        render_node(ctx, &mut scene.nodes, pts, node_id, &mut already_rendered)?;
     }
-    Ok(queue.into_iter())
+    Ok(())
 }
 
-fn enqueue_nodes(
-    parent_node_id: &NodeId,
-    queue: &mut Vec<NodeId>,
-    scene: &Scene,
+pub(super) fn render_node(
+    ctx: &mut RenderCtx,
+    nodes: &mut SceneNodesSet,
+    pts: Duration,
+    node_id: &NodeId,
+    already_rendered: &mut HashSet<NodeId>,
 ) -> Result<(), SceneError> {
-    let already_in_queue = queue.iter().any(|i| i == parent_node_id);
-    if already_in_queue {
+    if already_rendered.contains(node_id) {
         return Ok(());
     }
-    let parent_node = scene.nodes.node(parent_node_id)?;
-    for child in &parent_node.inputs {
-        enqueue_nodes(child, queue, scene)?;
+    // render all inputs
+    {
+        let input_ids: Vec<_> = nodes.node(node_id)?.inputs.to_vec();
+        for input_id in input_ids {
+            render_node(ctx, nodes, pts, &input_id, already_rendered)?;
+        }
+    }
+    // render node `node_id` if render had an effect fallback_id returned
+    // from the block will be set to None even if fallback is defined
```

Comment:
```suggestion
    // render node `node_id`. If render had an effect return `None`.
    // If rendering failed and the node has `fallback_id`, return `Some(fallback_id)`
```
Had a problem understanding what this means for a solid 10mins 😅